To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    # Can select any from the below:
    # "unsloth/Qwen2.5-0.5B", "unsloth/Qwen2.5-1.5B", "unsloth/Qwen2.5-3B"
    # "unsloth/Qwen2.5-14B",  "unsloth/Qwen2.5-32B",  "unsloth/Qwen2.5-72B",
    # And also all Instruct versions and Math. Coding verisons!
    model_name = "unsloth/Qwen2.5-7B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
Unsloth: Failed to patch Gemma3ForConditionalGeneration.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.3.19: Fast Qwen2 patching. Transformers: 4.51.1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors.index.json:   0%|          | 0.00/106k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.54G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/172 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/4.72k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/605 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/617 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [3]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.3.19 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [10]:
import logging # Optional: suppress warnings during map if needed
# logging.disable(logging.WARNING) # Uncomment this to suppress potential warnings from map

# Assume tokenizer is loaded and EOS_TOKEN is defined, for example:
# from transformers import AutoTokenizer
# tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1") # Or any relevant model
# EOS_TOKEN = tokenizer.eos_token
# If you don't have a tokenizer loaded yet, define EOS_TOKEN manually if you know it:
EOS_TOKEN = "</s>" # Example, replace with your actual EOS token if different

# Define the Alpaca prompt template
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{}

### Input:
{}

### Response:
{}"""

# FINAL CORRECTED formatting function using ACTUAL column names 'Question', 'Response'
def formatting_prompts_func(examples):
    # Use the ACTUAL column names from the dataset (CASE-SENSITIVE)
    questions = examples["Question"] # Access the 'Question' column
    responses = examples["Response"] # Access the 'Response' column
    # We will ignore 'Complex_CoT' for this formatting

    texts = []
    # Iterate using the correct variable names
    for question_text, response_text in zip(questions, responses):
        # Map dataset columns to the Alpaca template placeholders
        # 'Question' -> Instruction
        # No direct 'Input' column -> Leave empty
        # 'Response' -> Response
        text = alpaca_prompt.format(
            question_text, # Instruction field
            "",            # Input field (Leave empty as there's no separate input)
            response_text  # Response field
        ) + EOS_TOKEN # Must add EOS_TOKEN
        texts.append(text)

    # Return the dictionary with the new 'text' column
    return { "text" : texts }
# End of function definition

# Load the dataset
from datasets import load_dataset
dataset = load_dataset("Conard/fortune-telling", split = "train")

# --- Verification (as you ran before) ---
print("--- Verifying Dataset Columns Before Mapping ---")
print("Dataset Features:", dataset.features)
if dataset: # Check if dataset is not empty
    print("Keys in first example:", dataset[0].keys())
print("--- End Verification ---")
# --- End check ---

# Apply the mapping function using the corrected names
print("\nApplying the map function...")
dataset = dataset.map(formatting_prompts_func, batched = True,)
print("Map function applied successfully.")

# Optional: Verify the result
print("\nDataset processing complete.")
print("Example of formatted text:")
print(dataset[0]["text"])

# logging.disable(logging.NOTSET) # Uncomment to re-enable warnings if you disabled them

--- Verifying Dataset Columns Before Mapping ---
Dataset Features: {'Question': Value(dtype='string', id=None), 'Response': Value(dtype='string', id=None), 'Complex_CoT': Value(dtype='string', id=None)}
Keys in first example: dict_keys(['Question', 'Response', 'Complex_CoT'])
--- End Verification ---

Applying the map function...


Map:   0%|          | 0/207 [00:00<?, ? examples/s]

Map function applied successfully.

Dataset processing complete.
Example of formatted text:
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
新房装修,大门对着电梯好不好?要如何化解?

### Input:


### Response:
根据传统风水学的观点，大门正对电梯易形成"开口煞"，电梯频繁升降会扰乱家宅气场。建议化解方案：

1. 玄关阻隔法
在入门处设置L型屏风或文化砖玄关墙，高度以1.8米为宜，既保持采光又形成缓冲带

2. 五行通关法
门槛石下埋设五帝钱+白玉葫芦，建议选丙申年铸造的真品古币，配合门楣悬挂九宫八卦镜

3. 光影化解术
安装磨砂玻璃内推门，门框镶嵌黄铜门槛，每日辰时用海盐净化门廊区域

4. 现代科技方案
入户区安装智能感应灯带，设置循环播放的流水声效，运用声光电技术平衡磁场

需注意电梯井方位与家主命卦的关系，建议提供具体户型平面图进行吉凶方位测算。当代建筑中可采用半透明艺术隔断结合空气净化系统，既符合科学原理又兼顾传统智慧。</s>


<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [11]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        # num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/207 [00:00<?, ? examples/s]

In [12]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
7.266 GB of memory reserved.


In [13]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 207 | Num Epochs = 3 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 40,370,176/7,000,000,000 (0.58% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [14]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

756.3575 seconds used for training.
12.61 minutes used for training.
Peak reserved memory = 9.504 GB.
Peak reserved memory for training = 2.238 GB.
Peak reserved memory % of max memory = 64.473 %.
Peak reserved memory for training % of max memory = 15.182 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [23]:
# Assuming 'model', 'tokenizer', and 'FastLanguageModel' are already loaded
# Assuming 'alpaca_prompt' template is defined as before:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{}

### Input:
{}

### Response:
{}"""

# --- Snippet 1: Standard Generation ---

print("--- Running Standard Generation ---")
FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# Prepare the input using the CORRECT format based on the dataset
instruction_text = "命盤中破軍星在子女宮，對生育有影響嗎" # This comes from the 'Question' column
input_text = ""                                      # This field was empty during formatting
response_text = ""                                   # Leave blank for generation

inputs = tokenizer(
[
    alpaca_prompt.format(
        instruction_text, # Instruction field gets the actual question
        input_text,       # Input field is empty
        response_text     # Response field is empty for generation
    )
], return_tensors = "pt").to("cuda") # Ensure tensors are on the correct device

# Generate the output
outputs = model.generate(**inputs, max_new_tokens = 128, use_cache = True) # Increased max_new_tokens slightly

# Decode and print the result (including the prompt)
print("\nModel Output (Standard):")
print(tokenizer.batch_decode(outputs)[0]) # Decode the first element of the batch

# --- Snippet 2: Streaming Generation ---

print("\n--- Running Streaming Generation ---")
# No need to call FastLanguageModel.for_inference(model) again if already done

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer, skip_prompt=True) # skip_prompt=True to only see the generated part

# Prepare the input using the SAME CORRECT format
# (Using the same variables instruction_text, input_text, response_text from above)
inputs_streaming = tokenizer(
[
    alpaca_prompt.format(
        instruction_text, # Instruction field gets the actual question
        input_text,       # Input field is empty
        response_text     # Response field is empty for generation
    )
], return_tensors = "pt").to("cuda") # Ensure tensors are on the correct device

# Generate output using the streamer
print("\nModel Output (Streaming):")
_ = model.generate(**inputs_streaming, streamer = text_streamer, max_new_tokens = 128, use_cache = True) # use_cache is often beneficial
print("\n(Streaming finished)")

--- Running Standard Generation ---

Model Output (Standard):
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
命盤中破軍星在子女宮，對生育有影響嗎

### Input:


### Response:
破軍星在子女宮的命理解析

一、破軍星的本質特質
破軍屬火，為「將星」，主權柄、權力與決策能力。在子女宮中，它可能帶來以下特質：
1. 子女的教育與培養：破軍強調規範與紀律，子女可能需要嚴格的管教方式，但易形成過度保護或過度要求的教育模式。
2. 職業發展：子女可能在軍事、管理、法律等領域有較強適應力，但

--- Running Streaming Generation ---

Model Output (Streaming):
破軍星在子女宮的命理解析

一、破軍星的本質特質
破軍屬火，為「將星」，主權柄、權力與決策能力。在子女宮中，它可能帶來以下特質：
1. 子女的教育與培養：破軍強調規範與紀律，子女可能需要嚴格的管教方式，但易形成過度保護或過度要求的教育模式。
2. 職業發展：子女可能在軍事、管理、法律等領域有較強適應力，但

(Streaming finished)


 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [24]:
# Assuming 'model', 'tokenizer', and 'FastLanguageModel' are already loaded
# Assuming 'alpaca_prompt' template is defined as before:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{}

### Input:
{}

### Response:
{}"""

# Example Question from your screenshot
example_question = "庭院里种槐树对家运有什么影响?"

# --- Snippet 1: Standard Generation ---

print(f"--- Running Standard Generation for Question: '{example_question}' ---")
FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# Prepare the input using the CORRECT format based on the dataset structure
instruction_text = example_question # This comes from the 'Question' column
input_text = ""                     # Input field is empty as per our format
response_text = ""                  # Leave blank for generation

inputs = tokenizer(
[
    alpaca_prompt.format(
        instruction_text, # Instruction field gets the actual question
        input_text,       # Input field is empty
        response_text     # Response field is empty for generation
    )
], return_tensors = "pt").to("cuda") # Ensure tensors are on the correct device

# Generate the output - Adjust max_new_tokens based on expected output length
# Use a larger value if you expect outputs similar to Complex_CoT
outputs = model.generate(
    **inputs,
    max_new_tokens = 768, # Increased significantly for potentially long CoT outputs
    use_cache = True,
    # You might want to adjust other generation parameters like temperature, top_p etc.
    # temperature=0.7,
    # top_p=0.9,
)

# Decode and print the result (including the prompt)
print("\nModel Output (Standard):")
# Decoding the full output sequence
full_output = tokenizer.batch_decode(outputs)[0]
print(full_output)

# Optional: Extract just the generated part (after "### Response:")
response_marker = "### Response:"
response_start_index = full_output.find(response_marker)
if response_start_index != -1:
    generated_text = full_output[response_start_index + len(response_marker):].strip()
    print("\n--- Generated Text Only ---")
    print(generated_text)
    print("--- End Generated Text Only ---")


# --- Snippet 2: Streaming Generation ---

print(f"\n--- Running Streaming Generation for Question: '{example_question}' ---")
# No need to call FastLanguageModel.for_inference(model) again if already done

from transformers import TextStreamer
# skip_prompt=True is useful here to only see the generated part streamed
text_streamer = TextStreamer(tokenizer, skip_prompt=True)

# Prepare the input using the SAME CORRECT format
# (Using the same variables instruction_text, input_text, response_text from above)
inputs_streaming = tokenizer(
[
    alpaca_prompt.format(
        instruction_text, # Instruction field gets the actual question
        input_text,       # Input field is empty
        response_text     # Response field is empty for generation
    )
], return_tensors = "pt").to("cuda") # Ensure tensors are on the correct device

# Generate output using the streamer
print("\nModel Output (Streaming):")
_ = model.generate(
    **inputs_streaming,
    streamer = text_streamer,
    max_new_tokens = 768, # Keep max_new_tokens consistent
    use_cache = True,
    # Consistent generation parameters
    # temperature=0.7,
    # top_p=0.9,
)
print("\n(Streaming finished)")

--- Running Standard Generation for Question: '庭院里种槐树对家运有什么影响?' ---

Model Output (Standard):
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
庭院里种槐树对家运有什么影响?

### Input:


### Response:
**庭院槐树与家运的关联**  
槐树在传统文化中象征"富贵"与"长寿"，其树形挺拔、枝叶繁茂，常被用于庭院绿化。关于槐树对家运的影响，需结合具体方位、树龄、养护方式等综合判断：

### 1. **方位与五行**  
- **正东/东南方**：属木，主事业与健康，槐树宜种，可增强财运与活力。  
- **正西/西北方**：属金，若槐树枝叶茂盛，易形成"金木相生"之势，需注意修剪过度。  
- **正南/正北方**：属火/水，槐树不宜过密，以免形成"木旺克土"之象。

### 2. **树龄与形态**  
- **幼年期（3-5年）**：生命力旺盛，可增强宅主的事业心与进取心。  
- **中年期（6-10年）**：需注意病虫害，避免形成"木郁"之象。  
- **老年期（10年以上）**：枝干粗壮，易形成"木旺克土"之势，建议适当疏剪。

### 3. **养护要点**  
- **保持通风**：避免形成"木郁金"之象，可适当修剪侧枝。  
- **土壤透气**：使用沙质土壤，避免积水导致根部腐烂。  
- **季节管理**：春季施肥，秋季修剪，冬季防寒。

### 4. **禁忌与化解**  
- **忌与柏树同植**：柏树属金，易形成"金木相克"之象。  
- **忌在厨房附近**：槐树属木，易引发"火木相生"之灾。  
- **化解方法**：可在槐树下放置黄玉或黄水晶，增强木气。

**总结**：槐树在庭院中可增强宅主的事业心与活力，但需注意树形与方位的平衡。建议每年春季进行一次专业养护，可有效化解潜在的风水问题。具体布局需结合住宅坐向与家庭

<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# alpaca_prompt = You MUST copy from above!

inputs = tokenizer(
[
    alpaca_prompt.format(
        "What is a famous tall tower in Paris?", # instruction
        "", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "", # Get a token at https://huggingface.co/settings/tokens
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
